### PR TITLE
HPACK updates

### PR DESCRIPTION
**Pull request by ColdenCullen**
* Implement decode_string & get_encoded_length_[integer|string]
* Add support for using literal header values with indexed names

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/hpack.c [hunk 1]
```diff
@@ -255,48 +310,71 @@ void aws_hpack_context_destroy(struct aws_hpack_context *context) {
...
-    index -= s_static_header_table_size;
+    index -= STATIC_HEADER_TABLE_SIZE;
     AWS_ASSERT(index < context->dynamic_table.num_elements);
     return &context->dynamic_table
                 .buffer[(context->dynamic_table.index_0 + index) % context->dynamic_table.max_elements];
 }
 
-int aws_hpack_find_index(struct aws_hpack_context *context, const struct aws_http_header *header, size_t *index) {
+int aws_hpack_find_index(
+    struct aws_hpack_context *context,
```
Comment (graebm) [inline source/hpack.c]: trivial: const context?

File: include/aws/http/private/hpack.h [hunk 2]
```diff
@@ -35,16 +35,17 @@ void aws_hpack_static_table_clean_up(void);
...
-
+const struct aws_http_header *aws_hpack_get_header(struct aws_hpack_context *context, uint64_t index);
 AWS_HTTP_API
-int aws_hpack_find_index(struct aws_hpack_context *context, const struct aws_http_header *header, size_t *index);
-
+int aws_hpack_find_index(
+    struct aws_hpack_context *context,
+    const struct aws_http_header *header,
+    uint64_t *index,
+    bool *found_value);
```
Comment (graebm) [inline include/aws/http/private/hpack.h]: a little bit weird that the function has an int return type and also a `found_value` out param

does the int return type indicate catastrophic failure within the datastruture?

File: source/hpack.c [hunk 3]
```diff
@@ -255,48 +310,71 @@ void aws_hpack_context_destroy(struct aws_hpack_context *context) {
...
         if (absolute_index >= context->dynamic_table.index_0) {
             *index = absolute_index - context->dynamic_table.index_0;
         } else {
             *index = (context->dynamic_table.max_elements - context->dynamic_table.index_0) + absolute_index;
         }
         /* Need to add the static table size to re-base indicies */
-        *index += s_static_header_table_size;
+        *index += STATIC_HEADER_TABLE_SIZE;
         return AWS_OP_SUCCESS;
     }
```
Comment (graebm) [inline source/hpack.c]: Github won't let me comment on the `return AWS_OP_ERR;` 1 line down...

Are you sure that an aws_raise_error() occurs in the case that AWS_OP_ERR is returned?

File: source/hpack.c [hunk 4]
```diff
@@ -398,44 +506,126 @@ int aws_hpack_resize_dynamic_table(struct aws_hpack_context *context, size_t new
...
+    struct aws_byte_cursor *to_decode,
+    struct aws_byte_buf *output,
+    bool *huffman_encoded) {
+
+    AWS_PRECONDITION(context);
+    AWS_PRECONDITION(to_decode);
+    AWS_PRECONDITION(output);
+
+    if (!to_decode->len) {
+        return AWS_OP_ERR;
```
Comment (graebm) [inline source/hpack.c]: aws_raise_error(AWS_ERROR_INVALID_ARGUMENT)

File: source/hpack.c [hunk 5]
```diff
@@ -398,44 +506,126 @@ int aws_hpack_resize_dynamic_table(struct aws_hpack_context *context, size_t new
...
         int result = aws_huffman_encode(&context->encoder, to_encode, output);
         if (result) {
-            *to_encode = to_encode_backup;
-            return result;
+            goto error;
+        }
+    } else {
+        bool result = aws_byte_buf_write_from_whole_cursor(output, *to_encode);
+        if (!result) {
+            goto error;
```
Comment (graebm) [inline source/hpack.c]: aws_byte_buf_write_from_whole_cursor() doesn't raise an error

File: source/hpack.c [hunk 6]
```diff
@@ -398,44 +506,126 @@ int aws_hpack_resize_dynamic_table(struct aws_hpack_context *context, size_t new
...
+
+    bool use_huffman = *to_decode->ptr >> 7;
+    uint64_t value_length = 0;
+    if (aws_hpack_decode_integer(to_decode, 7, &value_length)) {
+        return AWS_OP_ERR;
+    }
+
+    struct aws_byte_cursor value = aws_byte_cursor_advance(to_decode, value_length);
+    if (!value.len) {
+        return AWS_OP_ERR;
```
Comment (graebm) [inline source/hpack.c]: aws_byte_cursor_advance() doesn't raise an error

File: source/hpack.c [hunk 7]
```diff
@@ -398,44 +506,126 @@ int aws_hpack_resize_dynamic_table(struct aws_hpack_context *context, size_t new
...
+    }
+
+    if (use_huffman) {
+        if (aws_huffman_decode(&context->decoder, &value, output)) {
+            return AWS_OP_ERR;
+        }
+    } else {
+        if (!aws_byte_buf_write_from_whole_cursor(output, value)) {
+            return AWS_OP_ERR;
```
Comment (graebm) [inline source/hpack.c]: byte_buf_write() functions don't raise an error

File: include/aws/http/private/hpack.h (hunk 2 above)
Comment (ColdenCullen) [reply include/aws/http/private/hpack.h]: `found_value` lets you know whether it found a value, or just a key associated with the header

File: source/hpack.c (hunk 1 above)
Comment (ColdenCullen) [reply source/hpack.c]: Not trivial, and can do!

File: source/hpack.c (hunk 3 above)
Comment (ColdenCullen) [reply source/hpack.c]: No, you make a good point. I'm going to refactor the signature of this function.

File: source/hpack.c (hunk 4 above)
Comment (ColdenCullen) [reply source/hpack.c]: SMORT

File: source/hpack.c [hunk 8]
```diff
@@ -246,6 +288,19 @@ struct aws_hpack_context *aws_hpack_context_new(struct aws_allocator *allocator,
...
+            &context->dynamic_table.reverse_lookup_name_only,
+            allocator,
+            max_dynamic_elements,
+            aws_hash_byte_cursor_ptr,
+            (aws_hash_callback_eq_fn *)aws_byte_cursor_eq,
+            NULL,
+            NULL)) {
+        aws_hash_table_clean_up(&context->dynamic_table.reverse_lookup);
+        aws_mem_release(allocator, context->dynamic_table.buffer);
+        return NULL;
```
Comment (graebm) [inline source/hpack.c]: your leaking the `context`. The if-statement above this one leaks it too
I suggest the usual `goto error` garbage

File: include/aws/http/private/hpack.h [hunk 9]
```diff
@@ -35,15 +35,16 @@ void aws_hpack_static_table_clean_up(void);
...
 AWS_HTTP_API
 void aws_hpack_context_destroy(struct aws_hpack_context *context);
 
 AWS_HTTP_API
-struct aws_http_header *aws_hpack_get_header(struct aws_hpack_context *context, size_t index);
-
+const struct aws_http_header *aws_hpack_get_header(const struct aws_hpack_context *context, uint64_t index);
 AWS_HTTP_API
-int aws_hpack_find_index(struct aws_hpack_context *context, const struct aws_http_header *header, size_t *index);
+uint64_t aws_hpack_find_index(
```
Comment (graebm) [inline include/aws/http/private/hpack.h]: comment that 0 means it wasn't found

**Pull request merged**